rust/src/trans/sum/float/mod.rs

```diff
@@ -0,0 +1,115 @@
+mod checked;
+use std::marker::PhantomData;
+
+pub use checked::*;
+
+mod ordered;
+pub use ordered::*;
+
+use num::{One, Zero};
+
+use crate::{
+    dist::IntDistance,
+    error::Fallible,
+    traits::{
+        AlertingAbs, CheckNull, ExactIntCast, FloatBits, InfAdd, InfCast, InfDiv, InfLog2, InfMul,
+        InfPow, InfSub, SaturatingAdd, TotalOrd,
+    },
+};
+
+use super::CanSumOverflow;
+
+pub trait Float:
+    CheckNull
+    + num::Bounded
+    + num::Float
+    + Clone
+    + TotalOrd
+    + ExactIntCast<usize>
+    + InfCast<IntDistance>
+    + InfAdd
+    + InfMul
+    + InfSub
+    + InfDiv
+    + InfPow
+    + InfLog2
+    + std::iter::Sum<Self>
+    + FloatBits
+    + ExactIntCast<Self::Bits>
+    + Zero
+    + One
+    + AlertingAbs
+    + SaturatingAdd
+    + CanSumOverflow
+{
+}
+impl<T> Float for T where
+    T: CheckNull
+        + num::Bounded
+        + num::Float
+        + Clone
+        + TotalOrd
+        + ExactIntCast<usize>
+        + InfCast<IntDistance>
+        + InfMul
+        + InfSub
+        + InfAdd
+        + InfDiv
+        + InfPow
+        + InfLog2
+        + std::iter::Sum<T>
+        + ExactIntCast<T::Bits>
+        + FloatBits
+        + Zero
+        + One
+        + AlertingAbs
+        + SaturatingAdd
+        + CanSumOverflow
+{
+}
+
+// Marker type to represent sequential, or recursive summation
+pub struct Sequential<T>(PhantomData<T>);
+
+// Marker type to represent pairwise, or cascading summation
+pub struct Pairwise<T>(PhantomData<T>);
```

Comment:
These marker types are really clever. Along with the SumRelaxation/UncheckedSum/SaturatingSum traits, it's a cool way to do dispatch on a single primitive type. I wonder if we could use this technique for other algorithmic variations?

Reply:
I really liked how it doesn't have an explicit argument, but it does have a downside. It ends up essentially becoming an explicit argument:
1. Rust functions don't support defaults for type args, so the type has to be inferred or set explicitly.
2. In this situation, the type can never be inferred from context, so the user has to figure out how to set it when using the rust API (import Pairwise, supply it as a type arg). I am able to set a default type arg in the python API, which I do.

Since it's now an explicit argument, it unfortunately also forces the user to specify the atomic type explicitly, or `_`.